### ISSUE TITLE
GINI_MEAN_DIFFERENCE Risk Measure for MeanRisk model [BUG]

### ISSUE THREAD

**Issue by sabirjana** [skfolio/skfolio#7]
GINI_MEAN_DIFFERENCE Risk Measure for MeanRisk model not working. I have tried with SCIP and CLARABEL solvers. Looks like I am missing something. Appreciate any help. Thanks!
```
model = MeanRisk(
    risk_measure=RiskMeasure.GINI_MEAN_DIFFERENCE,
    objective_function=ObjectiveFunction.MAXIMIZE_RATIO,
    portfolio_params=dict(name="Max Sharpe"), 
    solver='CLARABEL'
)
model.fit(X)
model.weights_-->
```

**Comment by HugoDelatte** [comment]
Your syntax is correct. Could you please provide the error message you are receiving?

The `GINI_MEAN_DIFFERENCE` risk measure can be slow to solve. Here are some tests I've run with a low-spec 8GB RAM Core i5 using the code below:

| Nb Observations  | CVXPY Compilation Time (s) | Total Time (s) |
| ------------------- | ----------------------- |--------------- |
| 100                       | 0.04                            | 0.7                 |
| 200                       | 0.1                              | 4.9                 |
| 300                       | 0.2                              | 20.4               |
| 400                       | 0.4                              | 40.5               |
| 1000                     | 4.0                              | 1018.9           |
| 2000                     | 21.7                            | X                    |


The time complexity seems to be around O(n<sup>3</sup>) with n being the number of observations. (For completeness, we should also analyze the complexity as a function of the number of assets).

Here are some tips when you are facing solver issues:
* Set `solver_params=dict(verbose=True)`
* Change the [solver parameter](https://www.cvxpy.org/tutorial/advanced/index.html#setting-solver-options) with `solver_params`. For example for CLARABEL: `solver_params={"tol_gap_abs": 1e-7, "tol_gap_rel": 1e-7}`
* Change solver, for example `solver="ECOS"`
* Change the scaling of the objective function and constraints with `scale_objective` and `scale_constraints` (useful when the solver doesn't automatically rescale the problem)

Here is the code I used:
```
from skfolio import RiskMeasure
from skfolio.datasets import load_sp500_dataset
from skfolio.optimization import MeanRisk, ObjectiveFunction
from skfolio.preprocessing import prices_to_returns

prices = load_sp500_dataset()
T = 100
X = prices_to_returns(prices.iloc[-T:])

model = MeanRisk(
    risk_measure=RiskMeasure.GINI_MEAN_DIFFERENCE,
    objective_function=ObjectiveFunction.MAXIMIZE_RATIO,
    solver_params=dict(verbose=True)
)
model.fit(X)
model.weights_
```

Let me add @phschiele, who works on CVXPY, in case he sees anything obvious to improve the time complexity (Gini optimization is done [here](https://github.com/skfolio/skfolio/blob/a1e79d3cf732467dcd078cd414b466b889a2b1c5/src/skfolio/optimization/convex/_base.py#L1904))

**Comment by sabirjana** [comment]
Thank you!! adding `solver_params `and `scale_objective, scale_constraints` helps however it is slow. You may want to check out implementation here if that helps to improve speed - https://github.com/dcajasn/Riskfolio-Lib/blob/master/examples/Tutorial%2037%20-%20OWA%20Portfolio%20Optimization.ipynb

**Comment by HugoDelatte** [comment]
I've compared it with Riskfolio; it's the same algorithm, with the only difference being the default parameters. For example, based on the above example and with 300 observations, they both take an average of 10.5s.

The parameters used for the comparison are:
* solver: CLARABEL
* solver params: {"tol_gap_abs": 1e-9, "tol_gap_rel": 1e-9}
* scale objective: 1000 (cannot be change in Riskfolio)
* scale constraints: 1000 (cannot be change in Riskfolio)

The Riskfolio code was based on the link you provided:
```
port = rp.Portfolio(returns=X)
method_mu='hist'
method_cov='hist'
port.assets_stats(method_mu=method_mu, method_cov=method_cov, d=0.94)
port.solvers = ['CLARABEL']
port.sol_params = {'CLARABEL': {"tol_gap_abs": 1e-9, "tol_gap_rel": 1e-9}}
model='Classic'
rm = 'GMD'
obj = 'Sharpe'
hist = True 
rf = 0 
l = 0 
w = port.optimization(model=model, rm=rm, obj=obj, rf=rf, l=l, hist=hist)
```

**Comment by sabirjana** [comment]
Thank you for the response! In Riskfolio when you use `rm='GMD'` with `owa `weights it faster.

```
owa_w = rp.owa_gmd(len(X))
w = port.owa_optimization(obj=obj,owa_w=owa_w,rf=rf,l=l)

```

**Comment by albystack** [comment]
Hi @HugoDelatte

I've now finished the investigation comparing the existing dense OWA formulation, constraint generation, and the p-norm approximation

The main conclusion is that I think the constraint-generation direction is worth pursuing further, with one correction to my original idea: the appropriate cuts are complete OWA permutation epigraph cuts rather than selected inequalities from the dense assignment dual

Using

$$
G(r)=\max_{P\in\mathcal P_T} a^TPr
$$

the separation problem can be solved by sorting the portfolio returns

By the rearrangement inequality this identifies the maximising permutation, and therefore the most violated epigraph cut, so the master problem only needs to retain the permutations actually encountered during optimisation

### Validation

I compared the method against the existing exact OWA formulation over 80 small/medium problems across:

- minimum risk
- maximum utility
- maximum ratio
- maximum-GMD constraints

All 80 CG runs converged, with a maximum final separation violation of around `1.0e-8` and a maximum relative objective difference from the dense formulation of `1.44e-6`

### Scaling

For `T=500, N=10`, median minimum-risk times were:

| Method | Time |
|---|---:|
| Existing dense OWA | 12.80 s |
| Constraint generation | 0.0358 s |
| p-norm | 0.1307 s |

The dense formulation timed out in every tested case with `T >= 1000`

At `T=5000`, CG took:

- 0.024 s for minimum risk
- 0.129 s for maximum ratio

compared with:

- 3.96 s for p-norm minimum risk
- 2.81 s for p-norm maximum ratio

I also tested the bundled S&P 500 data at `T=500, N=10`:

| Context | Dense OWA | CG | p-norm |
|---|---:|---:|---:|
| Minimum risk | 23.40 s | 0.107 s | 0.322 s |
| Maximum ratio | 23.59 s (`optimal_inaccurate`) | 0.089 s | failed |

### p-norm comparison

I reconstructed Cajas's p-norm formulation using the reference `(2,3,4,10,50)` tuple

It behaves as a conservative upper approximation of empirical GMD, rather than an exact replacement for it

Across the 80 optimised validation portfolios, represented-risk overestimation was:

- mean: **0.436%**
- median: **0.342%**
- 95th percentile: **0.944%**
- maximum: **1.511%**

The worst relative true-objective difference was **1.138%**, and the largest L1 distance between the p-norm and exact portfolio weights was **0.1008**

There also seems to be a solver trade-off, as the default p-norm formulation introduces `5T` 3D power cones

Clarabel handled the formulation in my environment, while the explicit formulation did not compile with SCIP/SciPy through CVXPY, and the default tuple was less robust in some maximum-ratio cases

### Trade-off

CG is not universally faster

For one `T=100` maximum-ratio case, dense OWA was slightly faster because CG needed around 50 cuts, and the number of cuts also increased with the number of assets

My current view is:

1. preserve the existing exact GMD semantics
2. retain dense OWA for sufficiently small problems
3. use permutation-cut constraint generation for larger problems
4. keep p-norm separate if an approximate GMD mode is ever desirable

I have the benchmark suit & raw results, and report as well if useful

Does this direction sound reasonable to you?

**Comment by HugoDelatte** [comment]
@albystack Very, very good analysis. So, as I understand it, the GC method is both faster and more accurate than the p-norm approximation.

The remaining question is about code complexity and modularity: can the GC method be implemented without compromising modularity and while keeping code complexity at an acceptable level?

Also, since the method is already fast on small problems, for the sake of code maintainability and cleanliness, I would probably avoid having two separate paths (exact GMD and GC). Using GC everywhere would simplify the code.

WDYT?

**Comment by albystack** [comment]
@HugoDelatte Yes, I agree. Using CG everywhere seems more preferable to maintaining 2 formulations. Especially since the occasional small problem slowdown appears quite modest

I think the implementation can remain modular by encapsulating the master/separation loop behind the already GMD interface: solve the master problem, sort the portfolio returns to generate the maximally violated permutation cut, and then repeat it until the separation tolerance is met

If the separation loop can stay fully encapsulated, I think we should be able to leave the public API and most of MeanRisk unchanged

Thoughts?

**Comment by HugoDelatte** [comment]
Perfect, sounds good to me. You can go ahead with the PR.
Thx!

**Comment by albystack** [comment]
Hi @HugoDelatte, so sorry for the wait. I’ve now opened PR #295 with the GMD constraint-generation implementation discussed above. Thanks again for the feedback and let me know what you think!
